datadog.go

```diff
@@ -0,0 +1,277 @@
+	return logrus.AllLevels[:h.MinLevel+1]
+}
+
+// Fire - implement Hook interface fire the entry
+func (h *DatadogHook) Fire(entry *logrus.Entry) error {
+	h.entryC <- *entry
+	return nil
+}
+
+func (h *DatadogHook) datadogURL() (string, error) {
+	u, err := url.Parse(h.datadogEndpoint + basePath)
+	if err != nil {
+		return "", err
+	}
+	parameters := url.Values{}
+	o := h.Tags
+	parameters.Add("ddsource", "golang")
+	parameters.Add("service", o.Service)
+	parameters.Add("hostname", o.Hostname)
+	var tags []string
+	tags = append(tags, fmt.Sprintf("environment:%s", o.Environment))
+	tags = append(tags, fmt.Sprintf("application:%s", o.Application))
+	tags = append(tags, fmt.Sprintf("service:%s", o.Service))
+	tags = append(tags, fmt.Sprintf("maintainer:%s", o.Maintainer))
+	tags = append(tags, fmt.Sprintf("reference:%s.%s.%s.%s", o.Maintainer, o.Application, o.Service, o.Environment))
+	parameters.Add("ddtags", strings.Join(tags, ","))
+	u.RawQuery = parameters.Encode()
+	return u.String(), nil
+}
+
+// This function listens to the log entry channel and the ticker channel and drives the batching of logs
+func (h *DatadogHook) batch(ticker <-chan time.Time) {
+	var batch [][]byte
+	size := 0
+	h.wg.Add(1)
+	go func() {
+		for entry := range h.entryC {
+			formatted, err := h.formatter.Format(&entry)
+			if err != nil {
+				h.errorC <- err
+				return
+			}
+
+			if size+len(formatted) >= maxContentSize || len(batch) == maxLogCount {
+				h.send(batch)
+				batch = make([][]byte, 0, maxLogCount)
+				size = 0
+			}
+
+			if len(formatted) > maxLogSize {
+				h.errorC <- fmt.Errorf("could not send log as it was too large! Maximum size for a single log is %d bytes, this log is %d bytes", maxLogSize, len(formatted))
+				return
+			}
+
+			batch = append(batch, formatted)
+			size += len(formatted)
+		}
+		h.wg.Done()
+	}()
+	h.wg.Add(1)
+	go func() {
+		for range ticker {
+			h.send(batch)
+			batch = make([][]byte, 0, maxLogCount)
+			size = 0
+		}
+		h.wg.Done()
+	}()
+}
+
+// Sends the Batch of logs to datadog
+func (h *DatadogHook) send(batch [][]byte) {
+	if h.ApiKey == "" || len(batch) == 0 {
+		return
+	}
+
+	buf := make([]byte, 0)
+	for i, line := range batch {
+		// First character is the opening bracket of the array
+		if i == 0 {
+			buf = append(buf, '[')
+		}
+		buf = append(buf, line...)
+		// Last character is the closing bracket of the array and doesn't get a ','
+		if i == len(batch)-1 {
+			buf = append(buf, ']')
+			continue
+		}
+		buf = append(buf, ',')
+	}
+
+	if len(buf) == 0 {
+		return
+	}
+
+	url, err := h.datadogURL()
+	if err != nil {
+		h.errorC <- err
+		return
+	}
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(buf))
+	if err != nil {
+		h.errorC <- err
+		return
+	}
+
+	header := http.Header{}
+	header.Add(apiKeyHeader, h.ApiKey)
+	header.Add("Content-Type", "application/json")
+	req.Header = header
+	i := 0
+	for {
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil || resp.StatusCode > 399 {
```

Comment:
This should be `err != nil && resp.StatusCode > 400` I think 🤔

Reply:
That would exclude 400 itself which is also an error we should retry is it not? Perhaps `>= 400` is more clear?

Reply:
I was getting confused I think, this is checking that there was an error, so its fine as is. 

Whichever way you want to do is fine by me, the logic is correct tho.